pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -24,307 +24,480 @@
     ARRAY_BASED,
-    MAP_BASED
+    LONG_MAP_BASED,
+    ARRAY_MAP_BASED
   }
 
   private final int _numGroupByColumns;
+  private final int[] _cardinalities;
+  private long _cardinalityProduct = 1L;
+  private final boolean[] _isSingleValueGroupByColumn;
+  private boolean _hasMultiValueGroupByColumn = false;
+  private final StorageType _storageType;
+  private int _numGroupKeys = 0;
+
   private final Dictionary[] _dictionaries;
+  // For single value columns.
   private final BlockValSet[] _singleBlockValSets;
+  // For multi value columns.
   private final BlockMultiValIterator[] _multiValIterators;
-  private final int[] _cardinalities;
-  private final boolean[] _isSingleValueGroupByColumn;
 
-  // Reusable arrays to avoid creating objects with each call.
+  // Reusable arrays for single value columns.
   private final int[][] _reusableSingleDictIds;
+  // Reusable buffer for multi value columns.
   private final int[] _reusableMultiValDictIdBuffer;
 
-  private StorageType _storageType;
-  private int _numUniqueGroupKeys;
-
   // For ARRAY_BASED storage type.
-  private boolean[] _uniqueGroupKeysFlag;
-
-  // For MAP_BASED storage type.
+  private boolean[] _groupKeyFlags;
+  // For LONG_MAP_BASED storage type.
   private Long2IntOpenHashMap _groupKeyToId;
+  // For ARRAY_MAP_BASED storage type.
+  private Object2IntOpenHashMap<IntArrayList> _arrayGroupKeyToId;
 
   /**
    * Constructor for the class. Initializes data members (reusable arrays).
    *
-   * @param dataFetcher
-   * @param groupByColumns
-   * @param maxNumGroupKeys
+   * @param dataFetcher data fetcher.
+   * @param groupByColumns group-by columns.
    */
-  DefaultGroupKeyGenerator(DataFetcher dataFetcher, String[] groupByColumns, int maxNumGroupKeys) {
+  DefaultGroupKeyGenerator(DataFetcher dataFetcher, String[] groupByColumns) {
     _numGroupByColumns = groupByColumns.length;
+    _cardinalities = new int[_numGroupByColumns];
+    _isSingleValueGroupByColumn = new boolean[_numGroupByColumns];
     _dictionaries = new Dictionary[_numGroupByColumns];
     _singleBlockValSets = new BlockValSet[_numGroupByColumns];
     _multiValIterators = new BlockMultiValIterator[_numGroupByColumns];
-    _cardinalities = new int[_numGroupByColumns];
-    _isSingleValueGroupByColumn = new boolean[_numGroupByColumns];
     _reusableSingleDictIds = new int[_numGroupByColumns][];
 
+    // Track the max number of values among all multi value group-by columns.
     int maxNumMultiValues = 0;
+
+    boolean longOverflow = false;
     for (int i = 0; i < _numGroupByColumns; i++) {
       DataSource dataSource = dataFetcher.getDataSourceForColumn(groupByColumns[i]);
-      _dictionaries[i] = dataSource.getDictionary();
-      _cardinalities[i] = dataSource.getDataSourceMetadata().cardinality();
-      _isSingleValueGroupByColumn[i] = dataSource.getDataSourceMetadata().isSingleValue();
 
+      // Store group-by column cardinalities and update cardinality product.
```

Comment:
I don't think that is necessary because the cardinality is associate with DataSource, and we need to maintain both cardinality product and long over flow flag. The current code logic is simple and if we extract a separate method, the code will become harder to understand